File: ios/Mercury/MercuryKit/Relay/RelayFraming.swift
import Foundation

/// Compact v1 plaintext records for bounded Noise application data.
///
/// Behavioral mirror of the plugin's `framing.py` and the frozen schema in
/// `protocol/schemas/framing-envelope-v1.json`: a fixed 50-byte header
/// (network byte order) followed by payload bytes, a 65,519-byte record cap,
/// a 1 MiB logical-message cap, canonical minimum fragmentation, and strictly
/// contiguous metadata-stable reassembly that resets on every rejection.
enum RelayFraming {
    static let magic: [UInt8] = [0x4D, 0x52] // "MR"
    static let protocolVersion: UInt8 = 1
    static let kindHermesBytes: UInt8 = 1
    static let channelIDSize = 16
    static let messageIDSize = 16
    static let headerSize = 50
    static let maxNoisePlaintextBytes = 65_519
    static let maxPayloadBytes = maxNoisePlaintextBytes - headerSize // 65,469
    /// Raised from the initial 1 MiB so attachments comparable to direct
    /// mode fit through the relay; kept under the router's 25 MB / 10 s
    /// per-socket byte budget so one maximal message cannot trip it. Must
    /// match the plugin's `framing.MAX_LOGICAL_MESSAGE_BYTES`.
    static let maxLogicalMessageBytes = 16 << 20
    static let maxFragmentCount = (maxLogicalMessageBytes + maxPayloadBytes - 1) / maxPayloadBytes

    struct DecodeError: Error, Equatable {
        let reason: String
    }

    struct Frame: Equatable {
        let channelID: Data
        let messageID: Data
        let fragmentIndex: Int
        let fragmentCount: Int
        let logicalLength: Int
        let payload: Data
    }

    static func canonicalFragmentCount(logicalLength: Int) -> Int {
        max(1, (logicalLength + maxPayloadBytes - 1) / maxPayloadBytes)
    }

    private static func canonicalPayloadLength(logicalLength: Int, fragmentIndex: Int) -> Int {
        min(maxPayloadBytes, logicalLength - fragmentIndex * maxPayloadBytes)
    }

    /// Encode one bounded logical byte string into canonical records.
    static func encodeMessage(
        channelID: Data,
        messageID: Data,
        payload: Data
    ) throws -> [Data] {
        guard channelID.count == channelIDSize, messageID.count == messageIDSize else {
            throw DecodeError(reason: "invalid identifier length")
        }
        guard payload.count <= maxLogicalMessageBytes else {
            throw DecodeError(reason: "logical length exceeds v1 limit")
        }
        let count = canonicalFragmentCount(logicalLength: payload.count)
        var records: [Data] = []
        records.reserveCapacity(count)
        for index in 0..<count {
            let start = payload.index(payload.startIndex, offsetBy: index * maxPayloadBytes)
            let length = canonicalPayloadLength(
                logicalLength: payload.count, fragmentIndex: index
            )
            let end = payload.index(start, offsetBy: length)
            var record = Data(capacity: headerSize + length)
            record.append(contentsOf: magic)
            record.append(protocolVersion)
            record.append(kindHermesBytes)
            record.append(0) // flags
            record.append(0) // reserved
            appendUInt16(&record, UInt16(index))
            appendUInt16(&record, UInt16(count))
            appendUInt32(&record, UInt32(payload.count))
            appendUInt32(&record, UInt32(length))
            record.append(channelID)
            record.append(messageID)
            record.append(payload[start..<end])
            records.append(record)
        }
        return records
    }

    /// Validate and decode one complete canonical plaintext record.
    static func decodeRecord(_ record: Data) throws -> Frame {
        guard record.count >= headerSize else {
            throw DecodeError(reason: "truncated record header")
        }
        guard record.count <= maxNoisePlaintextBytes else {
            throw DecodeError(reason: "record exceeds Noise plaintext limit")
        }
        let bytes = [UInt8](record)
        guard bytes[0] == magic[0], bytes[1] == magic[1] else {
            throw DecodeError(reason: "invalid magic")
        }
        guard bytes[2] == protocolVersion else { throw DecodeError(reason: "unknown version") }
        guard bytes[3] == kindHermesBytes else { throw DecodeError(reason: "unknown kind") }
        guard bytes[4] == 0 else { throw DecodeError(reason: "unknown flags") }
        guard bytes[5] == 0 else { throw DecodeError(reason: "reserved value is nonzero") }
        let fragmentIndex = Int(readUInt16(bytes, 6))
        let fragmentCount = Int(readUInt16(bytes, 8))
        let logicalLength = Int(readUInt32(bytes, 10))
        let payloadLength = Int(readUInt32(bytes, 14))
        guard fragmentCount >= 1, fragmentCount <= maxFragmentCount else {
            throw DecodeError(reason: "fragment count is outside the v1 range")
        }
        guard logicalLength <= maxLogicalMessageBytes else {
            throw DecodeError(reason: "logical length exceeds v1 limit")
        }
        let expectedCount = canonicalFragmentCount(logicalLength: logicalLength)
        guard fragmentIndex >= 0, fragmentIndex < expectedCount else {
            throw DecodeError(reason: "fragment index or length is not canonical")
        }
        guard fragmentCount == expectedCount else {
            throw DecodeError(reason: "fragment count is not canonical")
        }
        let expectedPayloadLength = canonicalPayloadLength(
            logicalLength: logicalLength, fragmentIndex: fragmentIndex
        )
        guard payloadLength == expectedPayloadLength else {
            throw DecodeError(reason: "payload length is not canonical")
        }
        guard record.count == headerSize + payloadLength else {
            throw DecodeError(reason: "record length does not match payload length")
        }
        return Frame(
            channelID: Data(bytes[18..<34]),
            messageID: Data(bytes[34..<50]),
            fragmentIndex: fragmentIndex,
            fragmentCount: fragmentCount,
            logicalLength: logicalLength,
            payload: Data(bytes[headerSize...])
        )
    }

    private static func appendUInt16(_ data: inout Data, _ value: UInt16) {
        data.append(UInt8(value >> 8))
        data.append(UInt8(value & 0xFF))
    }

    private static func appendUInt32(_ data: inout Data, _ value: UInt32) {
        data.append(UInt8((value >> 24) & 0xFF))
        data.append(UInt8((value >> 16) & 0xFF))
        data.append(UInt8((value >> 8) & 0xFF))
        data.append(UInt8(value & 0xFF))
    }

    private static func readUInt16(_ bytes: [UInt8], _ offset: Int) -> UInt16 {
        UInt16(bytes[offset]) << 8 | UInt16(bytes[offset + 1])
    }

    private static func readUInt32(_ bytes: [UInt8], _ offset: Int) -> UInt32 {
        UInt32(bytes[offset]) << 24 | UInt32(bytes[offset + 1]) << 16
            | UInt32(bytes[offset + 2]) << 8 | UInt32(bytes[offset + 3])
    }
}

/// Reassemble one logical message from strictly ordered, matching records.
/// Any rejection resets the in-progress message before rethrowing.
final class RelayFrameReassembler {
    private let expectedChannelID: Data?
    private var messageID: Data?
    private var activeChannelID: Data?
    private var fragmentCount = 0
    private var logicalLength = 0
    private var nextIndex = 0
    private var receivedLength = 0
    private var parts: [Data] = []

    init(channelID: Data? = nil) {
        expectedChannelID = channelID
    }

    var inProgress: Bool { messageID != nil }

    func reset() {
        messageID = nil
        activeChannelID = nil
        fragmentCount = 0
        logicalLength = 0
        nextIndex = 0
        receivedLength = 0
        parts = []
    }

    /// Consume one record; returns the complete logical message when the
    /// final canonical fragment arrives, nil while in progress.
    func push(_ record: Data) throws -> Data? {
        do {
            let frame = try RelayFraming.decodeRecord(record)
            if let expectedChannelID, frame.channelID != expectedChannelID {
                throw RelayFraming.DecodeError(reason: "metadata channel mismatch")
            }
            if messageID == nil {
                guard frame.fragmentIndex == 0 else {
                    throw RelayFraming.DecodeError(reason: "order requires fragment zero first")
                }
                activeChannelID = frame.channelID
                messageID = frame.messageID
                fragmentCount = frame.fragmentCount
                logicalLength = frame.logicalLength
            } else if frame.channelID != activeChannelID
                || frame.messageID != messageID
                || frame.fragmentCount != fragmentCount
                || frame.logicalLength != logicalLength {
                throw RelayFraming.DecodeError(
                    reason: "metadata channel, message, or length mismatch"
                )
            }
            guard frame.fragmentIndex == nextIndex else {
                throw RelayFraming.DecodeError(reason: "order is not contiguous")
            }
            guard receivedLength + frame.payload.count <= logicalLength else {
                throw RelayFraming.DecodeError(reason: "received length exceeds logical length")
            }
            parts.append(frame.payload)
            receivedLength += frame.payload.count
            nextIndex += 1
            guard nextIndex == fragmentCount else { return nil }
            guard receivedLength == logicalLength else {
                throw RelayFraming.DecodeError(reason: "received length is incomplete")
            }
            var message = Data(capacity: receivedLength)
            for part in parts { message.append(part) }
            reset()
            return message
        } catch {
            reset()
            throw error
        }
    }
}

File: shared/mercury-core/src/commonMain/kotlin/com/unsupportedpastels/mercury/core/relay/RelayFraming.kt
package com.unsupportedpastels.mercury.core.relay

internal class RelayFramingException(
    val reason: String,
) : Exception("Mercury Relay frame validation failed")

class RelayFrame(
    val channelId: ByteArray,
    val messageId: ByteArray,
    val fragmentIndex: Int,
    val fragmentCount: Int,
    val logicalLength: Int,
    val payload: ByteArray,
)

object RelayFraming {
    val magic = byteArrayOf(0x4d, 0x52)
    const val protocolVersion = 1
    const val kindHermesBytes = 1
    const val channelIdSize = 16
    const val messageIdSize = 16
    const val headerSize = 50
    const val maxNoisePlaintextBytes = 65_519
    const val maxPayloadBytes = maxNoisePlaintextBytes - headerSize
    const val maxLogicalMessageBytes = 16 shl 20
    const val maxFragmentCount = (maxLogicalMessageBytes + maxPayloadBytes - 1) / maxPayloadBytes

    fun canonicalFragmentCount(logicalLength: Int): Int =
        maxOf(1, (logicalLength + maxPayloadBytes - 1) / maxPayloadBytes)

    @Throws(RelayFramingException::class)
    fun encodeMessage(channelId: ByteArray, messageId: ByteArray, payload: ByteArray): List<ByteArray> {
        if (channelId.size != channelIdSize || messageId.size != messageIdSize) fail("invalid identifier length")
        if (payload.size > maxLogicalMessageBytes) fail("logical length exceeds v1 limit")
        val count = canonicalFragmentCount(payload.size)
        return List(count) { index ->
            val start = index * maxPayloadBytes
            val length = canonicalPayloadLength(payload.size, index)
            ByteArray(headerSize + length).also { record ->
                record[0] = magic[0]
                record[1] = magic[1]
                record[2] = protocolVersion.toByte()
                record[3] = kindHermesBytes.toByte()
                record[4] = 0
                record[5] = 0
                writeUInt16(record, 6, index)
                writeUInt16(record, 8, count)
                writeUInt32(record, 10, payload.size)
                writeUInt32(record, 14, length)
                channelId.copyInto(record, 18)
                messageId.copyInto(record, 34)
                payload.copyInto(record, headerSize, start, start + length)
            }
        }
    }

    @Throws(RelayFramingException::class)
    fun decodeRecord(record: ByteArray): RelayFrame {
        if (record.size < headerSize) fail("truncated record header")
        if (record.size > maxNoisePlaintextBytes) fail("record exceeds Noise plaintext limit")
        if (record[0] != magic[0] || record[1] != magic[1]) fail("invalid magic")
        if (record[2].toInt() and 0xff != protocolVersion) fail("unknown version")
        if (record[3].toInt() and 0xff != kindHermesBytes) fail("unknown kind")
        if (record[4].toInt() != 0) fail("unknown flags")
        if (record[5].toInt() != 0) fail("reserved value is nonzero")
        val fragmentIndex = readUInt16(record, 6)
        val fragmentCount = readUInt16(record, 8)
        val logicalLength = readUInt32(record, 10)
        val payloadLength = readUInt32(record, 14)
        if (fragmentCount !in 1..maxFragmentCount) fail("fragment count is outside the v1 range")
        if (logicalLength !in 0..maxLogicalMessageBytes) fail("logical length exceeds v1 limit")
        val expectedCount = canonicalFragmentCount(logicalLength)
        if (fragmentIndex !in 0 until expectedCount) fail("fragment index or length is not canonical")
        if (fragmentCount != expectedCount) fail("fragment count is not canonical")
        val expectedLength = canonicalPayloadLength(logicalLength, fragmentIndex)
        if (payloadLength != expectedLength) fail("payload length is not canonical")
        if (record.size != headerSize + payloadLength) fail("record length does not match payload length")
        return RelayFrame(
            channelId = record.copyOfRange(18, 34),
            messageId = record.copyOfRange(34, 50),
            fragmentIndex = fragmentIndex,
            fragmentCount = fragmentCount,
            logicalLength = logicalLength,
            payload = record.copyOfRange(headerSize, record.size),
        )
    }

    fun decodeFailureReason(record: ByteArray): String? =
        try {
            decodeRecord(record)
            null
        } catch (error: RelayFramingException) {
            error.reason
        }

    private fun canonicalPayloadLength(logicalLength: Int, fragmentIndex: Int): Int =
        minOf(maxPayloadBytes, logicalLength - fragmentIndex * maxPayloadBytes)

    private fun writeUInt16(bytes: ByteArray, offset: Int, value: Int) {
        bytes[offset] = (value ushr 8).toByte()
        bytes[offset + 1] = value.toByte()
    }

    private fun writeUInt32(bytes: ByteArray, offset: Int, value: Int) {
        bytes[offset] = (value ushr 24).toByte()
        bytes[offset + 1] = (value ushr 16).toByte()
        bytes[offset + 2] = (value ushr 8).toByte()
        bytes[offset + 3] = value.toByte()
    }

    private fun readUInt16(bytes: ByteArray, offset: Int): Int =
        (bytes[offset].toInt() and 0xff shl 8) or (bytes[offset + 1].toInt() and 0xff)

    private fun readUInt32(bytes: ByteArray, offset: Int): Int =
        (bytes[offset].toInt() and 0xff shl 24) or
            (bytes[offset + 1].toInt() and 0xff shl 16) or
            (bytes[offset + 2].toInt() and 0xff shl 8) or
            (bytes[offset + 3].toInt() and 0xff)

    private fun fail(reason: String): Nothing = throw RelayFramingException(reason)
}

class RelayFrameReassembler(
    private val expectedChannelId: ByteArray? = null,
) {
    private var messageId: ByteArray? = null
    private var activeChannelId: ByteArray? = null
    private var fragmentCount = 0
    private var logicalLength = 0
    private var nextIndex = 0
    private var receivedLength = 0
    private var parts = mutableListOf<ByteArray>()

    val inProgress: Boolean get() = messageId != null

    fun reset() {
        messageId = null
        activeChannelId = null
        fragmentCount = 0
        logicalLength = 0
        nextIndex = 0
        receivedLength = 0
        parts = mutableListOf()
    }

    @Throws(RelayFramingException::class)
    fun push(record: ByteArray): ByteArray? {
        try {
            val frame = RelayFraming.decodeRecord(record)
            if (expectedChannelId != null && !frame.channelId.contentEquals(expectedChannelId)) {
                fail("metadata channel mismatch")
            }
            if (messageId == null) {
                if (frame.fragmentIndex != 0) fail("order requires fragment zero first")
                activeChannelId = frame.channelId
                messageId = frame.messageId
                fragmentCount = frame.fragmentCount
                logicalLength = frame.logicalLength
            } else if (!frame.channelId.contentEquals(activeChannelId!!) ||
                !frame.messageId.contentEquals(messageId!!) || frame.fragmentCount != fragmentCount ||
                frame.logicalLength != logicalLength
            ) {
                fail("metadata channel, message, or length mismatch")
            }
            if (frame.fragmentIndex != nextIndex) fail("order is not contiguous")
            if (receivedLength + frame.payload.size > logicalLength) fail("received length exceeds logical length")
            parts += frame.payload
            receivedLength += frame.payload.size
            nextIndex += 1
            if (nextIndex != fragmentCount) return null
            if (receivedLength != logicalLength) fail("received length is incomplete")
            val result = ByteArray(receivedLength)
            var offset = 0
            parts.forEach { part ->
                part.copyInto(result, offset)
                offset += part.size
            }
            reset()
            return result
        } catch (error: Throwable) {
            reset()
            throw error
        }
    }

    private fun fail(reason: String): Nothing = throw RelayFramingException(reason)
}
